event_model/documents/event_descriptor.py

```diff
@@ -0,0 +1,129 @@
+from typing import TYPE_CHECKING, Any, Dict, List
+
+from typing_extensions import Annotated, NotRequired, Literal, TypedDict
+
+from ._type_wrapper import Field
+
+
+class DataKey(TypedDict):
+    external: NotRequired[
+        Annotated[
+            str,
+            Field(
+                description="Where the data is stored if it is stored external to the events.",
+                regex=r"^[A-Z]+:?",
+            ),
+        ]
+    ]
+    dims: NotRequired[
+        Annotated[
+            List[str],
+            Field(
+                description="The names for dimensions of the data. Null or empty list if scalar data",
+            ),
+        ]
+    ]
+    object_name: NotRequired[
+        Annotated[
+            str,
+            Field(description="The name of the object this key was pulled from."),
+        ]
+    ]
+
+    dtype: Annotated[
+        Literal["string", "number", "array", "boolean", "integer"],
+        Field(description="The type of the data in the event."),
+    ]
+
+    shape: Annotated[
+        List[int],
+        Field(description="The shape of the data.  Empty list indicates scalar data."),
+    ]
+    source: Annotated[
+        str, Field(description="The source (ex piece of hardware) of the data.")
+    ]
+
+
+if TYPE_CHECKING:
+    ObjectHints = Any
+else:
+
+    class ObjectHints(TypedDict):
+        __root__: Annotated[Any, Field(title="Object Hints")]
```

Comment:
As discussed in person, this should be `Annotated[Any, AsRef("ObjectHints")]` and the creation of a `BaseModel` class with `__root__` annotation can happen in the schema creation